### ISSUE TITLE
Remove (don't load) models that are a single node

### ISSUE THREAD

**Issue by ukemi** [geneontology/pathways2GO#202]
In some cases, a Reactome pathway doesn't have any reactions that are directly associated with it. Instead it has a collection of subpathways under it. In those cases, the parent gets imported as a single node with nothing else associated with it. We should not load these.
eg R-HSA-71291

**Comment by deustp01** [comment]
Here's a weedy suggestion for how to proceed. Agreed that a pathway with no content except subpathways yields a GO-CAM with no informative content, but before simply discarding these it would be prudent to get a list of all the single-node pathways for manual inspection to confirm that nothing is lost. I expect that everything on the list will be OK for removal - even where a curator has made one of these as a placeholder and plan to fill in individual reaction children along with the pathway children, when and if that happens the pathway will then pass the rule proposed here and will get loaded OK.

And a naive question. Is the Reactome event hierarchy somehow preserved in the exported GO-CAM structure? I guess that it is not and in that case, these empty grouping pathways do not have a useful linking role.

**Comment by ukemi** [comment]
> And a naive question. Is the Reactome event hierarchy somehow preserved in the exported GO-CAM structure? I guess that it is not and in that case, these empty grouping pathways do not have a useful linking role.

This is a great question that has now set me thinking. The Reactome hierarchy is not preserved because of the inability to discriminate is_a and Part_of (another thing that I think we could brainstorm about at a face-2-face. I think you had some good ideas about this). However, let's say that there is a Reactome pathway that has no reactions, but only pathways as children. If the parent pathway has an asserted GO BP term mapped and none of the children do, it would be safe to put the parent pathway on generic children. It doesn't matter is the child is a subclass or a part of the parent because we won't represent that. The parent BP will just go to the new top node of the model. I'm not sure how many of these exist, but I think I've seen some.

**Comment by deustp01** [comment]
Now deferred from things to do in connection with GO-CAM build from Reactome 82 - make this a headache for another day

**Comment by ukemi** [comment]
QA for @ukemi and @deustp01. Once this is done, it should only eliminate the grouping pathways that have no reactions associated with them.

**Comment by dustine32** [comment]
I'm trying to knock this ticket out as part of getting the comprehensive list of "done" Reactome GO-CAMs with no molecular event placeholder activities. When applying the criteria "contains no molecular events," I kept seeing these empty parent pathway models containing only a single BP node in the results. Simply blocking the write-out of these model files would clean up these results.

So I applied this filter and did a before-after comparison to get the list of models (351 total) that would be removed, attached below:
[pathways_dropped_no_activities.txt](https://github.com/user-attachments/files/17218316/pathways_dropped_no_activities.txt)

@ukemi @deustp01 Please take a look and let me know if you catch anything wrong with this list. Thanks!

**Comment by ukemi** [comment]
@deustp01

**Comment by deustp01** [comment]
To track checking, made a Google Doc, ["Getting rid of models that are a single node" ](https://docs.google.com/document/d/1nQMrdu7zM70V2yj9AyWt9-Ob5_jti6MkMIf1L53OLZ4/edit)in the "Getting rid of molecular events" folder.

**Comment by deustp01** [comment]
I've now checked the list.

Bottom line - 350 pathways should be blocked as Dustin proposes and one should be edited out of existence in Reactome.

Weedy details -

Most of the pathways on the ["Getting rid of models that are a single node" ](https://docs.google.com/document/d/1nQMrdu7zM70V2yj9AyWt9-Ob5_jti6MkMIf1L53OLZ4/edit) list have only one or more other pathways as children (indicated with a simple “no” in column 2 in the list / table). While they provide useful grouping information for a future GO-CAM structure that shows causal relationships among pathways, blocking their write-out now is correct.

A subset of these have only a single pathway as a child (“no” in column 2, “has only one pathway child” in column 3). Dustin should block these just like the simple-no ones. They are flagged because in Reactome as in GO it does not make much sense for a higher-level grouping term to have only a single child, so a useful side-effect of this review of candidates for blocking is a list of pathways in Reactome that are candidates for rearrangement / merging to eliminate unneeded steps in the Reactome pathway hierarchy.

One pathway, R-HSA-1630316, contains a single reaction child but this can be fixed on the Reactome side by putting the reaction in a different pathway, which actually it belongs to.

A number of pathways are composed of multiple drug (one of the participants is flagged as a drug) or stealth-drug (one of the participants is a set all of whose members are drugs) reactions. The GO-CAM script correctly suppresses the generation of activity units from these reactions, resulting in an empty pathway. These too should be blocked at the write-out stage. I’ve flagged them separately (“YES” in column 2 and drug verbiage in column 3) because I guess that Dustin may want to reinforce the say-no-to-drugs tests to filter them out more elegantly?

**Comment by ukemi** [comment]
>  While they provide useful grouping information for a future GO-CAM structure that shows causal relationships among pathways, blocking their write-out now is correct.

This is indeed true. One task to make use of this in GO-CAM would be to try to figure out a method for determining if the child pathways are is_a or part_of children of the grouping pathway. Right now it is a mix.

**Comment by deustp01** [comment]
> figure out a method for determining if the child pathways are is_a or part_of children of the grouping pathway.

While we've been aware of the distinction, we never had to deal with it within the Reactome event hierarchy, so now there's a big legacy clean-up problem. My hunch is that it would be straightforward for a human curator to make the classification correctly, but tedious. @dustine32 do you see any hope here for a script that could sort out the two classes of pathway reliably, and identify pathways that have both is_a and part_of children, because I expect we have some - there's nothing to prohibit them?

**Comment by dustine32** [comment]
> do you see any hope here for a script that could sort out the two classes of pathway reliably, and identify pathways that have both is_a and part_of children

Sorry @ukemi @deustp01, I just noticed this request as I came to announce I was ready to merge the single-node pathway fix code. For the two aspects of this request:
1. These two classes of pathway (I assume, the single-node, zero activity pathways) would be (1) contains only a single child pathway and (2) contains multiple pathway children, right?
2. Discerning pathway hierarchy relationships of `is_a` vs `part_of` - I do not see any distinction of these relations in the BioPAX file nor on the Reactome site. Could you point me to wherever you're seeing these relationship types and I can try tracing from there?

Also, I think this still means I'm good to start merging the single-node pathway filtering code into `master`. I'll at least open a PR.

**Comment by ukemi** [comment]
This might be something worth looking at in New York. I have looked before and discriminating between is_a and Part_of was not obvious. However, one thing I didn't do was look at preceding reaction relationships. Another possibility is to look at the asserted BPs that are on the pathways and interrogate the ontology for relationships, but I'm not sure this will work. I think there are a lot of 'partial' pathways that are asserted to be the pathway. We might want to have a look at some of these in NYC. They are not asserted, so you won't see them in the BioPax. 
Here is an example:
All of the subpathways of "Signaling by Receptor Tyrosine Kinases" (R-HSA-9006934) are subtypes of the parent.
But the children of "Signaling by EGFR" (R-HSA-177929) are almost all part_of.

**Comment by ukemi** [comment]
@dustine32 Small sample, but just a sanity check. Don't know what's going on with 2066762. Do we close if these work on dev or do we wait until they go to production?
Tested : Single node models
R-ALL-2066762-- not found in production, not in dev
R-HSA-8979227-- found in production, not in dev
R-HSA-9006934-- found in production, not in dev
R-HSA-190236-- found in production, not in dev
R-HSA-109582-- found in production, not in dev

Positive control: Models with reactions
R-HSA-418346-- found in production, found in dev
R-HSA-75105-- found in production, found in dev
R-HSA-211945-- found in production, found in dev
R-HSA-71384-- found in production, found in dev
R-HSA-1268020-- found in production, found in dev

**Comment by dustine32** [comment]
@ukemi Thank you for checking! R-ALL-2066762 appears to be a small mol [1-3-oxo-THA-CoA](https://reactome.org/content/detail/R-ALL-2066762#Homo%20sapiens), not a pathway. Was this in a report I generated?

So, other than R-ALL-2066762, your results are pretty much what we expected.

> Do we close if these work on dev or do we wait until they go to production?

I suppose the more correct thing is to wait until it's in prod but that's up to how we want to manage this project. For sure, I think this ticket can get moved out of "Software in progress", perhaps to "Release pipeline run in progress"?

As for classifying pathway hierarchy relations/groupings, it would be much better to spin out a new ticket since the basic "drop single node pathways" aspect has been addressed.

**Comment by ukemi** [comment]
@dustine32  Sorry, looks like that ID was a cut and paste error on my part. It was a chemical in the pathway on which I was working. But now that you remind me, I think you did make a report that @deustp01 was also looking at. Let's go ahead and move this over as you suggest unless @deustp01 can remember any reasons why we shouldn't.

I agree that we should split out the other issue about the is_a versus part_of hierarchy. That is something we have thought about for a while but have not made a lot of progress on. At the end of the day we don't want these single node pathways no matter what. They don't have any value in their current state.
